### PR TITLE
test(core): verify Unicode and ICU versions cross-platform 🙀 

### PR DESCRIPTION
**Pull request by srl295**
- load versions from node and ICU4C++ and Blocks.txt
- validate Node version against package.json's engine
- for now, don't complain if ICU4C++ mismatches, but DO check Unicode versions.

Fixes: #10183

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by srl295** [comment]
Example output

```
= test_all
== load_json loading tests/unit/ldml/nodeversions.json
== load_json loading ../../../../package.json
== test_unicode_versions
cxx_icu	73.1
cxx_icu_unicode	15.0
node_icu_unicode	15.0
node	18.16.0
node_icu	72.1
node_engine	^18.x
block_unicode_ver	15.1.0.txt
```

**Comment by srl295** [comment]
So linux is using ICU from the environment, which only supports Unicode 14 not 15.

```
  stdout:
20:08:09   = test_all
20:08:09   == load_json loading tests/unit/ldml/nodeversions.json
20:08:09   == load_json loading ../../../../package.json
20:08:09   == test_unicode_versions
20:08:09   cxx_icu  70.1
20:08:09   cxx_icu_unicode  14.0
20:08:09   node_icu_unicode  15.0
20:08:09   node  18.17.1
20:08:09   node_icu  73.1
20:08:09   node_engine  ^18.x
20:08:09   block_unicode_ver  15.1.0.txt
20:08:09   stderr:
20:08:09   Test failed at ../../../tests/unit/ldml/test_unicode.cpp:110:
20:08:09   expected: 14
20:08:09   actual:   15
```

**Comment by srl295** [comment]
> So linux is using ICU from the environment, which only supports Unicode 14 not 15.
> 
> ```
>   stdout:
> 20:08:09   = test_all
> 20:08:09   == load_json loading tests/unit/ldml/nodeversions.json
> 20:08:09   == load_json loading ../../../../package.json
> 20:08:09   == test_unicode_versions
> 20:08:09   cxx_icu  70.1
> 20:08:09   cxx_icu_unicode  14.0
> 20:08:09   node_icu_unicode  15.0
> 20:08:09   node  18.17.1
> 20:08:09   node_icu  73.1
> 20:08:09   node_engine  ^18.x
> 20:08:09   block_unicode_ver  15.1.0.txt
> 20:08:09   stderr:
> 20:08:09   Test failed at ../../../tests/unit/ldml/test_unicode.cpp:110:
> 20:08:09   expected: 14
> 20:08:09   actual:   15
> ```

I'm going to take out the assert for the cxx_icu version, for now.

**Comment by srl295** [comment]
all working but wasm

**Comment by srl295** [comment]
OK ptal, web unrelated?

File: core/tests/unit/ldml/meson.build [hunk 1]
```diff
@@ -35,20 +35,46 @@ if node.found()
   subdir('invalid-keyboards')
 endif
 
-
 # Build ldml test executable
 
+keyboard_build_path = join_paths(meson.current_build_dir(),'keyboards')
```
Comment (mcdurdin) [inline core/tests/unit/ldml/meson.build:40]:
```suggestion
keyboard_build_path = meson.current_build_dir() / 'keyboards'
```

It's cleaner to use `/` now in meson path construction

File: core/tests/unit/ldml/meson.build [hunk 2]
```diff
@@ -94,14 +120,34 @@ if cpp_compiler.get_id() == 'emscripten'
   normalization_tests_flags += ['-lnodefs.js', '-sEXPORTED_RUNTIME_METHODS=[\'UTF8ToString\']']
 endif
 
-t = executable('test_context_normalization',
+tc = executable('test_context_normalization',
```
Comment (mcdurdin) [inline core/tests/unit/ldml/meson.build:123]:
```suggestion
test_context_normalization = executable('test_context_normalization',
```

Can we use the full process name as the variable name? Makes future maintenance so much easier

File: core/tests/unit/ldml/meson.build [hunk 3]
```diff
@@ -94,14 +120,34 @@ if cpp_compiler.get_id() == 'emscripten'
...
     include_directories: [inc, libsrc, '../../../../developer/src/ext/json'],
     link_args: links + normalization_tests_flags,
     dependencies: [icu_uc, icu_i18n],
     objects: lib.extract_all_objects(recursive: false))
-test('test_context_normalization', t, suite: 'ldml')
+test('test_context_normalization', tc, suite: 'ldml')
+
+# Build and run additional test_unicode test
+
+u = executable('test_unicode', 'test_unicode.cpp',
```
Comment (mcdurdin) [inline core/tests/unit/ldml/meson.build:134]:
```suggestion
test_unicode = executable('test_unicode', 'test_unicode.cpp',
```

Ditto

**Review by mcdurdin — CHANGES_REQUESTED** [review]
Looks fine, just small nits

**Review by mcdurdin — APPROVED** [review]

**Comment by mcdurdin** [comment]
@ermshiperete the API Verification build is failing, any ideas why? Log is inconclusive.

**Comment by ermshiperete** [comment]
> @ermshiperete the API Verification build is failing, any ideas why? Log is inconclusive.

Ignore the API Verification for now. I thought it would work, but I must be overlooking something and it's still not working. I'll add a step to always greenbar until that's solved. Sorry for the noise.

**Comment by ermshiperete** [comment]
> Example output
> 
> ```
> = test_all
> == load_json loading tests/unit/ldml/nodeversions.json
> == load_json loading ../../../../package.json
> == test_unicode_versions
> cxx_icu	73.1
> cxx_icu_unicode	15.0
> node_icu_unicode	15.0
> node	18.16.0
> node_icu	72.1
> node_engine	^18.x
> block_unicode_ver	15.1.0.txt
> ```

I'm wondering if it would be better to change the order of the output so that all ICU versions are together, all Unicode versions are together etc. ? That way it would be easier to spot which versions should be the same.

**Comment by srl295** [comment]
> > Example output
> > ```
> > = test_all
> > == load_json loading tests/unit/ldml/nodeversions.json
> > == load_json loading ../../../../package.json
> > == test_unicode_versions
> > cxx_icu	73.1
> > cxx_icu_unicode	15.0
> > node_icu_unicode	15.0
> > node	18.16.0
> > node_icu	72.1
> > node_engine	^18.x
> > block_unicode_ver	15.1.0.txt
> > ```
> 
> I'm wondering if it would be better to change the order of the output so that all ICU versions are together, all Unicode versions are together etc. ? That way it would be easier to spot which versions should be the same.

Good idea. Maybe with some explanatory text, even.

I think I'll do this as a separate PR since this stack is getting longer.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by srl295** [comment]
> > Example output
> > ```
> > = test_all
> > == load_json loading tests/unit/ldml/nodeversions.json
> > == load_json loading ../../../../package.json
> > == test_unicode_versions
> > cxx_icu	73.1
> > cxx_icu_unicode	15.0
> > node_icu_unicode	15.0
> > node	18.16.0
> > node_icu	72.1
> > node_engine	^18.x
> > block_unicode_ver	15.1.0.txt
> > ```
> 
> I'm wondering if it would be better to change the order of the output so that all ICU versions are together, all Unicode versions are together etc. ? That way it would be easier to spot which versions should be the same.

https://github.com/keymanapp/keyman/pull/11572

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Pull request merged**